### PR TITLE
Allow build to work behind corporate proxy servers

### PR DESCRIPTION
**Pull request by jlanng**
Many corporate proxies require default user credentials to be sent. This PR allows the CLI to be downloaded and the build to run to completion in those environments

### CODE REVIEW

**Comment by joperezr** [comment]
Hey @jlanng, I'm trying to understand why this is the only place where we need to make a change, and not with our other network dependencies. For example, we download a lot of nuget packages, so if you have to authenticate I would expect that to fail as well with 407.

**Comment by joperezr** [comment]
cc: @weshaggard

**Comment by joperezr** [comment]
Personally I feel like this is a very specific scenario for a very small number of people, so I'm not sure how I feel about putting the logic in such a general place. Is there some sort of environment that you can setup for powershell before calling build.cmd in order for us not having to do this and still work with proxies? If so, I think I would rather have the instructions for that in a wiki.

That said, I might not be understanding the full spectrum of the problem, so please feel free to give more info on this.

**Comment by weshaggard** [comment]
I tend to agree with @joperezr I'm all for supporting proxy scenarios but I suspect there are likely other places that need this in order to work. We have https://github.com/dotnet/corefx/issues/15160 tracking the more general problem. @jlanng do you know if there are other issues blocking the scenario?

**Comment by jlanng** [comment]
@joperezr nuget appears to understand how to handle 407s on Windows.  The build runs to completion with this PR.  I have to say that don't accept your assertion that this only affects a small number of people. Do you have stats to support that? Most places I've worked (Microsoft shops) have had a similar proxy setup. I wonder how many potential contributors have tried to build the repo and given up? 

@weshaggard 
I do see that this is an isolated fix for a general problem. The long Powershell script on a single line already captures/dupes retry logic and is quite hard to read. Perhaps we could have an additional scriptlet just for downloading files, which can then capture all the edge cases? If we need to have duplication, perhaps it's better to be a whole file that can easily be diffed?

**Comment by weshaggard** [comment]
@jlanng do you know if setting HTTPS_PROXY one windows works or does the .NET APIs not honor that? If not I wonder if there is another windows tool that would honor it so we can use that for both our linux and windows builds to enable the proxy scenario.

If we can make that work that would be the best route to go but if not then I think we can work on updating the script to enable it.

**Comment by jlanng** [comment]
@weshaggard no, sadly not.  Only apps ported from *nix support that, e.g. curl and git.

If we get this merged, I could wait a couple of weeks to ensure it doesn't cause any issues, then work through the other builds to ensure they all work.

**Comment by davidsh** [comment]
@weshaggard 
In terms of .NET Core, the Windows implementation already looks by default in the "system" settings (i.e. Internet Explorer registry settings) to find a proxy.  Not sure what .NET Core *Nix does (CurlHandler).  

The issue you are running into here in this PR is solving the credentials problem where the proxy wants creds.  Normally, if this happens, most LAN based proxy servers on a corporate network will allow the use of default logged-on credentials.  And that requires having the proxy server request creds in Negotiate or NTLM.  Basic auth, for example, can't be used to transmit default logged-on creds.  Explicit creds can always be used as well.

So, if the proxy can use default credentials, then they need to be specified to the proxy.  And the "Default" proxy creds can be specified either via DefaultWebProxy or use HttpClient.DefaultProxyCredentials.

So, bottom line, code changes like this PR are required.  There are no settings to store the credentials in HTTPS_PROXY or other registry settings.

**Comment by weshaggard** [comment]
Thanks @davidsh.

Thanks @jlanng for fixing the script. I've filed https://github.com/dotnet/buildtools/issues/1648 to track trying to improve this for all our repo's and putting it into a shared script of some sort.
